### ISSUE TITLE
"Kennfeld" is different for each heatpump model

### ISSUE THREAD

**Issue by MadOne** [OStrama/weishaupt_modbus#18]
**Is your feature request related to a problem? Please describe.**
The different heatpump models have different "Kennfelder"

**Describe the solution you'd like**
I'd like to autodetect the heatpump model and select the correct "Kennfeld"

**Describe alternatives you've considered**
Select the correct heatpump model in setup.

**Additional context**
[PR for WSB8](https://github.com/OStrama/weishaupt_modbus/pulls)

**Comment by MadOne** [comment]
I dont see any model info in the modbus doku.
The hostname of my heatpump is: WWP-CPU-COM

Any suggestions on detecting the heatpump model are welcome.

**Comment by OStrama** [comment]
I also do not see any info about it in the known Modbus entries. However, in WEM-Portal expert mode I see this:

![image](https://github.com/user-attachments/assets/f6dbec1a-2bd4-49b1-8d2e-2351361162e3)

I think there are modbus addresses there that we do not know...


On the other hand, this could also be something that the user can configure at installation.

**Comment by OStrama** [comment]
Another point:

I did my first tries for interpolation with scipy CubicSpline. The resultet curves look perfectly like the ones in my WBB 12 manual:

![image](https://github.com/user-attachments/assets/bceba496-132b-4952-9669-02ce5d027c25)

However, since in some HA implementations it seems to be an issue to use scipy, I switched to Chebyceff interpolation from numpy. The result is not as good:

![image](https://github.com/user-attachments/assets/14d9544e-a1b5-44c5-886f-ee650b678c7f)

If there's any idea how to install scipy reliably, it would make power estimation much better

**Comment by MadOne** [comment]
Is there no way to export the equation?

**Comment by MadOne** [comment]
-I was already thinking of scanning all possible (don't know yet) registers for a valid response. So we could find out undocumented information.
-We could reach out to Weishaupt for the most up to date documentation / information

**Comment by edlestau** [comment]
Hi MadOne.

If you contact Weisshaupt, you could perhaps also ask why some data via Modbus are less accurate than the comparable data via the portal. An example are the statistical data on electricity consumption and the amount of energy generated.

In my case, for example, the power consumption for the day is displayed as
 8.945 KWh at the Portal and  8.0 kWh at the Modbus.

It would also be interesting to know whether and where the long-term and medium-term values for the outdoor temperature can be read out via Modbus.
These can be found in the portal, but according to the Modbus interface documentation, they cannot be read out.

Thanks foryour great work!
edle_stau

**Comment by MadOne** [comment]
@edlestau 
As far is i understand modbus can only provide integers. The integer part of 8.945 is 8. It would be clever to round it up but sadly this is what they did.
At temperatures the multiply with 10. So 30,0°C become 300. We devide this by 10 and and get 30.0°C.
This could have been done with the Statistics values too, but sadly the diddnt. Im quite sure the never will, because this would break all existing code.

They could use a propper rounding and this would half the possible error.

@OStrama :
What do you think about the suggestions above?
Another option would be to ship scipy with our integration and import it locally.

**Comment by OStrama** [comment]
Hmm, I do not want to implement splines in Phyton from scratch (https://de.wikipedia.org/wiki/Spline)... 
The Scipy library is written in C/C++, so it has to be compiled for every platform. Some of them are available by simply installing it via PIP, on some even not (the HA box). But Numpy is. As long as HA does not decide to deliver it, I think we have no chance. The quwestion is: What will really be the difference in real live. +/- 0.1 kW would not be an issue at the end..

**Comment by MadOne** [comment]
I was [playing around](https://github.com/OStrama/weishaupt_modbus/blob/main/custom_components/weishaupt_modbus/cubic_splines_playground.py) a bit with scipy and cubic splines. 

We could get the equations or just calc all possible values.

**Comment by OStrama** [comment]
Yes, taking the json files, calc all samples and store it in the json file would be a solution

the polynominal coefficients are calculated to fit to this certain curve by the spline algorithm. Thea are different for every curve..

**Comment by MadOne** [comment]
Jeah we would have to calculate them for every curve. Just like the values.

**Comment by MadOne** [comment]
We added a dropdown for known values.
Will close this here.
